staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -0,0 +1,619 @@
+// returns nil in the case that there is no available concurrency or
+// the queuelengthlimit has been reached
+// NOTE: This function must only be called with the QueueSet locked
+func (qs *queueSetImpl) timeoutOldRequestsAndRejectOrEnqueue(hashValue uint64) *fq.Request {
+
+	//	Start with the shuffle sharding, to pick a queue.
+	queueIdx := qs.ChooseQueueIdx(hashValue)
+	queue := qs.queues[queueIdx]
+	// The next step is the logic to reject requests that have been waiting too long
+	qs.removeTimedOutRequestsFromQueue(queue)
+	// NOTE: currently timeout is only checked for each new request.  This means that there can be
+	// requests that are in the queue longer than the timeout if there are no new requests
+	// We think this is a fine tradeoff
+
+	// Create a request and enqueue
+	req := &fq.Request{
+		DequeueChannel: make(chan bool, 1),
+		EnqueueTime:    qs.clk.Now(),
+		Queue:          queue,
+	}
+	if ok := qs.rejectOrEnqueue(req); !ok {
+		return nil
+	}
+	metrics.ObserveQueueLength(qs.name, len(queue.Requests))
+	return req
+
+}
+
+// removeTimedOutRequestsFromQueue rejects old requests that have been enqueued
+// past the requestWaitLimit
+func (qs *queueSetImpl) removeTimedOutRequestsFromQueue(queue *fq.Queue) {
+	timeoutIdx := -1
+	now := qs.clk.Now()
+	pkts := queue.Requests
+	// pkts are sorted oldest -> newest
+	// can short circuit loop (break) if oldest packets are not timing out
+	// as newer packets also will not have timed out
+
+	// now - requestWaitLimit = waitLimit
+	waitLimit := now.Add(-qs.requestWaitLimit)
+	for i, pkt := range pkts {
+		if waitLimit.After(pkt.EnqueueTime) {
+			qs.counter.Add(1)
+			pkt.DequeueChannel <- false
+			close(pkt.DequeueChannel)
+			// // TODO(aaron-prindle) verify this makes sense here
+			// get idx for timed out packets
+			timeoutIdx = i
+
+		} else {
+			break
+		}
+	}
+	// remove timed out packets from queue
+	if timeoutIdx != -1 {
+		// timeoutIdx + 1 to remove the last timeout pkt
+		removeIdx := timeoutIdx + 1
+
+		// remove all the timeout packets
+		queue.Requests = pkts[removeIdx:]
+		// decrement the # of requestsEnqueued
+		qs.numRequestsEnqueued -= removeIdx
+	}
+}
+
+// GetRequestsExecuting gets the # of requests which are "executing":
+// this is the# of requests/packets which have been dequeued but have not had
+// finished (via the FinishRequest method invoked after service)
+func (qs *queueSetImpl) GetRequestsExecuting() int {
+	total := 0
+	for _, queue := range qs.queues {
+		total += queue.RequestsExecuting
+	}
+	return total
+}
+
+// ChooseQueueIdx uses shuffle sharding to select a queue index
+// using a 'hashValue'.  The 'hashValue' derives a hand from a set range of
+// indexes (range 'desiredNumQueues') and returns the queue with the least queued packets
+// from a dealt hand (of size 'handSize')
+func (qs *queueSetImpl) ChooseQueueIdx(hashValue uint64) int {
+	// TODO(aaron-prindle) currently a lock is held for this in a larger anonymous function
```

Comment:
This comment also needs updating.  The lock is acquired on entry to Wait.

Reply:
Updated comment to read:
`	// A lock is held for this in Wait
`